### PR TITLE
Refactor/intellij version classes

### PR DESCRIPTION
**Pull request by MaciejG604**

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -275,10 +275,10 @@ For instance:
 1. our plugin in version `0.7.0` is compatible with IntelliJ `2020.2`
 2. then IntelliJ `2020.3-EAP` is released (see [snapshot repository](https://www.jetbrains.com/intellij-repository/snapshots/) -> Ctrl+F `idea`)
 3. we check if `0.7.0` is compatible with IntelliJ `2020.3-EAP`:
-   set `mapInstance["eapOfLatestSupportedMajor"]` in [IntellijVersionHelper](buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersionHelper.kt)
+   set `eapOfLatestSupportedMajor` in [IntellijVersionHelper](buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersionHelper.kt)
```
Comment (PawelLipski) [inline CONTRIBUTING.md:278]: In `IntellijVersions`, right?

File: build.gradle.kts [hunk 2]
```diff
@@ -238,15 +235,15 @@ tasks.withType<RunIdeTask> { maxHeapSize = "4G" }
 
 tasks.withType<RunPluginVerifierTask> {
   val maybeEap =
-      if (intellijVersions["eapOfLatestSupportedMajor"] != null)
+      if (IntellijVersions.eapOfLatestSupportedMajor != null)
           listOf(
-              (intellijVersions["eapOfLatestSupportedMajor"] as String).replace(
+              (IntellijVersions.eapOfLatestSupportedMajor!!).replace(
```
Comment (PawelLipski) [inline build.gradle.kts:240]: Is `!!` needed here? I suppose Kotlin can infer that at this point the value cannot be null

File: build.gradle.kts [hunk 3]
```diff
@@ -238,15 +235,15 @@ tasks.withType<RunIdeTask> { maxHeapSize = "4G" }
...
                   "-EAP-(CANDIDATE-)?SNAPSHOT".toRegex(), ""))
       else emptyList()
   ideVersions.set(
       listOf(
-          *(intellijVersions["latestMinorsOfOldSupportedMajors"] as List<String>).toTypedArray(),
-          intellijVersions["latestStable"] as String,
+          *(IntellijVersions.latestMinorsOfOldSupportedMajors).toTypedArray(),
```
Comment (PawelLipski) [inline build.gradle.kts:245]: Nit: hmm maybe join the lists/elements with some `+` operator or whatever is available in Kotlin? I suppose a cast to an array isn't the most elegant solution here

File: buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersionHelper.kt [hunk 4]
```diff
@@ -52,11 +20,20 @@ object IntellijVersionHelper {
...
-    val versionValue = instance[versionKey] ?: return emptyList()
-    if (versionValue is List<*>) {
-      return versionValue as List<String>
+    return when (versionKey) {
+      "earliestSupportedMajor" -> listOf(IntellijVersions.earliestSupportedMajor)
+      "latestMinorsOfOldSupportedMajors" -> IntellijVersions.latestMinorsOfOldSupportedMajors
+      "latestStable" -> listOf(IntellijVersions.latestStable)
+      "latestSupportedMajor" -> listOf(IntellijVersions.latestSupportedMajor)
+      "buildTarget" -> listOf(IntellijVersions.buildTarget)
+      "eapOfLatestSupportedMajor" -> {
```
Comment (PawelLipski) [inline buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersionHelper.kt:29]: Or https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/list-of-not-null.html

File: buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersions.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package com.virtuslab.gitmachete.buildsrc
+
+// See https://www.jetbrains.com/intellij-repository/releases/ -> Ctrl+F .idea
+object IntellijVersions {
+
+    val intellijVersionsProp = IntellijVersionHelper.getProperties()
```
Comment (PawelLipski) [inline buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersions.kt:6]: `private`?

File: buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersions.kt [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    // 211.4961.33-EAP-SNAPSHOT)
+    // rather than a "rolling snapshot" X-EAP-SNAPSHOT (e.g. 211-EAP-SNAPSHOT)
+    // to ensure that the builds are reproducible.
+    // EAP-CANDIDATE-SNAPSHOTs can be used for binary compatibility checks,
+    // but for some reason aren't resolved in UI tests.
+    // Generally, see https://www.jetbrains.com/intellij-repository/snapshots/ -> Ctrl+F .idea
+    // Use `null` if the latest supported major has a stable release (and not just EAPs).
+    val eapOfLatestSupportedMajor: String? = intellijVersionsProp.getProperty("eapOfLatestSupportedMajor")
+
+    val latestSupportedMajor = if (eapOfLatestSupportedMajor != null)
```
Comment (PawelLipski) [inline buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersions.kt:27]: `: String`

File: buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersions.kt (hunk 5 above)
Comment (MaciejG604) [reply buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersions.kt:6]: Wasn't there a problem with private vals in Kotlin? Or was that something else?

File: buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/UpdateEapBuildNumber.kt [hunk 7]
```diff
@@ -45,13 +45,10 @@ open class UpdateEapBuildNumber : DefaultTask() {
   @TaskAction
   fun execute() {
     val properties = IntellijVersionHelper.getProperties()
-    val latestEapBuildNumber =
-        IntellijVersionHelper.instance["eapOfLatestSupportedMajor"] as String?
+    val latestEapBuildNumber = IntellijVersions.eapOfLatestSupportedMajor
```
Comment (PawelLipski) [inline buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/UpdateEapBuildNumber.kt:48]: Nit: inline this val

File: buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersions.kt (hunk 5 above)
Comment (PawelLipski) [reply buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersions.kt:6]: Huh there is, I forbade them with a regex xDD although this doesn't matter as much in build code, you can probably exclude them from `scripts/prohibit-kotlin-private-fields`

File: build.gradle.kts (hunk 2 above)
Comment (MaciejG604) [reply build.gradle.kts:240]: Apparently the compiler can deal only with simple statements :(

Comment (PawelLipski) [reply build.gradle.kts:240]: Checker Framework (which is use for null safety in Java) is then much smarter 🧠

File: buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/UpdateEapBuildNumber.kt (hunk 7 above)
Comment (MaciejG604) [reply buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/UpdateEapBuildNumber.kt:48]: Is it possible in Kotlin?

Comment (PawelLipski) [reply buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/UpdateEapBuildNumber.kt:48]: As in:

File: buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/UpdateEapBuildNumber.kt [hunk 8]
```diff
@@ -45,13 +45,10 @@ open class UpdateEapBuildNumber : DefaultTask() {
...
 
-    val buildNumberThreshold =
-        if (!latestEapBuildNumber.isNullOrEmpty()) latestEapBuildNumber.replace("-EAP-SNAPSHOT", "")
-        else
-            "${IntellijVersionHelper.toBuildNumber(IntellijVersionHelper.instance["latestStable"] as String)}.999999.999999"
+    val buildNumberThreshold = latestEapBuildNumber?.replace("-EAP-SNAPSHOT", "") ?:
```
Comment (PawelLipski) [inline buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/UpdateEapBuildNumber.kt:50]:
```suggestion
    val buildNumberThreshold = IntellijVersions.eapOfLatestSupportedMajor?.replace("-EAP-SNAPSHOT", "") ?:
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -275,10 +275,10 @@ For instance:
 1. our plugin in version `0.7.0` is compatible with IntelliJ `2020.2`
 2. then IntelliJ `2020.3-EAP` is released (see [snapshot repository](https://www.jetbrains.com/intellij-repository/snapshots/) -> Ctrl+F `idea`)
 3. we check if `0.7.0` is compatible with IntelliJ `2020.3-EAP`:
-   set `mapInstance["eapOfLatestSupportedMajor"]` in [IntellijVersionHelper](buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersionHelper.kt)
+   set `eapOfLatestSupportedMajor` in [IntellijVersions](buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersionHelper.kt)
```
Comment (PawelLipski) [inline CONTRIBUTING.md:278]: Plus update the target file as well!

**Review by PawelLipski — APPROVED** [review]

File: buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/UpdateEapBuildNumber.kt [hunk 10]
```diff
@@ -47,7 +47,7 @@ open class UpdateEapBuildNumber : DefaultTask() {
     val properties = IntellijVersionHelper.getProperties()
     val latestEapBuildNumber = IntellijVersions.eapOfLatestSupportedMajor
```
Comment (PawelLipski) [inline buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/UpdateEapBuildNumber.kt:48]: Aaand remove this one now (btw, Ctrl+Alt+N to inline a value, at least on Linux)

**Review by PawelLipski — APPROVED** [review]

**Pull request merged**